mmdet3d/ops/pointnet_modules/point_sa_module.py

```diff
@@ -144,44 +185,116 @@ def forward(
                 Index of the features.
         """
         new_features_list = []
-        xyz_flipped = points_xyz.transpose(1, 2).contiguous()
-        if indices is not None:
-            assert (indices.shape[1] == self.num_point[0])
-            new_xyz = gather_points(xyz_flipped, indices).transpose(
-                1, 2).contiguous() if self.num_point is not None else None
-        elif target_xyz is not None:
-            new_xyz = target_xyz.contiguous()
-        else:
-            indices = self.points_sampler(points_xyz, features)
-            new_xyz = gather_points(xyz_flipped, indices).transpose(
-                1, 2).contiguous() if self.num_point is not None else None
+
+        # sample points, (B, num_point, 3), (B, num_point)
+        new_xyz, indices = self._sample_points(points_xyz, features, indices,
+                                               target_xyz)
 
         for i in range(len(self.groupers)):
-            # (B, C, num_point, nsample)
-            new_features = self.groupers[i](points_xyz, new_xyz, features)
+            # grouped_results may contain:
+            # - grouped_features: (B, C, num_point, nsample)
+            # - grouped_xyz: (B, 3, num_point, nsample)
+            # - grouped_idx: (B, num_point, nsample)
+            grouped_results = self.groupers[i](points_xyz, new_xyz, features)
 
             # (B, mlp[-1], num_point, nsample)
-            new_features = self.mlps[i](new_features)
-            if self.pool_mod == 'max':
-                # (B, mlp[-1], num_point, 1)
-                new_features = F.max_pool2d(
-                    new_features, kernel_size=[1, new_features.size(3)])
-            elif self.pool_mod == 'avg':
-                # (B, mlp[-1], num_point, 1)
-                new_features = F.avg_pool2d(
-                    new_features, kernel_size=[1, new_features.size(3)])
-            else:
-                raise NotImplementedError
+            new_features = self.mlps[i](grouped_results)
 
-            new_features = new_features.squeeze(-1)  # (B, mlp[-1], num_point)
+            # this is a bit hack because PAConv outputs two values
+            # we take the first one as feature
+            if isinstance(self.mlps[i][0], PAConv):
+                assert isinstance(new_features, tuple)
+                new_features = new_features[0]
+
+            # (B, mlp[-1], num_point)
+            new_features = self._pool_features(new_features)
             new_features_list.append(new_features)
 
         return new_xyz, torch.cat(new_features_list, dim=1), indices
 
 
+@SA_MODULES.register_module()
```

Comment:
we can use `@SA_MODULES.register_module(name='PointSAMSG')` to simplify the name

Reply:
This will introduce a BC-breaking as some other models also use SA modules. Do you think we need to do this?